File: JavaImplementation/LineSweep.java
import java.util.*;

public class LineSweep {
    private SortedList lineSweepStatus;
    private Queue<Event> eventQueue;
    public Set<Point> intersections;

    public LineSweep(ArrayList<Edge> edges) {
        this.lineSweepStatus = new SortedList();
        this.eventQueue = new PriorityQueue<>(new event_comparator());
        ArrayList<Event> events = getEvents(edges);
        for (Event e : events) {
            eventQueue.add(e);
        }
        this.intersections = new HashSet<Point>();
    }

    public void line_sweep() {
        while (!eventQueue.isEmpty()) {
            Event currEvent = eventQueue.poll();
            double currY = currEvent.point.y;
            if (currEvent.category == 0) {

                lineSweepStatus.add(currEvent.edge1, currY);
                Edge predEdge = lineSweepStatus.predecessor(currEvent.edge1);
                Edge sucEdge = lineSweepStatus.successor(currEvent.edge1);
                if (predEdge != null) {
                    Point intersection = predEdge.intersectionPoint(currEvent.edge1);
                    this.addIntersection(intersection, predEdge, currEvent.edge1);
                }
                if (sucEdge != null) {
                    Point intersection = sucEdge.intersectionPoint(currEvent.edge1);
                    this.addIntersection(intersection, currEvent.edge1, sucEdge);
                }

            } else if (currEvent.category == 1) {
                Edge predEdge = lineSweepStatus.predecessor(currEvent.edge1);
                Edge sucEdge = lineSweepStatus.successor(currEvent.edge1);
                if (predEdge != null && sucEdge != null) {
                    Point intersection = predEdge.intersectionPoint(sucEdge);
                    this.addIntersection(intersection, predEdge, sucEdge);
                }
                lineSweepStatus.remove(currEvent.edge1);

            } else {

                lineSweepStatus.swapEdges2(currEvent.edge1, currEvent.edge2, currY);

                    Edge predEdge = lineSweepStatus.predecessor(currEvent.edge2);
                    if (predEdge != null) {
                        Point intersection = predEdge.intersectionPoint(currEvent.edge2);
                        addIntersection(intersection, predEdge, currEvent.edge2);
                    }
                    Edge sucEdge = lineSweepStatus.successor(currEvent.edge1);
                    if (sucEdge != null) {
                        Point intersection = currEvent.edge1.intersectionPoint(sucEdge);
                        addIntersection(intersection, currEvent.edge1, sucEdge);
                    }

            }
        }
    }

    public void addIntersection(Point intersection, Edge edge1, Edge edge2) {
        if (intersection != null) {
            if (!intersections.contains(intersection)) {
                this.intersections.add(intersection);
                this.eventQueue.add(new Event(2, intersection, edge1, edge2));

            }
        }
    }


    public ArrayList<Event> getEvents(ArrayList<Edge> edges) {
        // Event category 0 is bottom, 1 is top, 2 is intersection
        ArrayList<Event> events = new ArrayList<Event>();
        for (Edge edge : edges) {
            if (edge.p0.y > edge.p1.y) {
                events.add(new Event(1, edge.p0, edge, null));
                events.add(new Event(0, edge.p1, edge, null));
            } else {
                events.add(new Event(0, edge.p0, edge, null));
                events.add(new Event(1, edge.p1, edge, null));
            }
        }
        return events;
    }

    private class event_comparator implements Comparator<Event> {
        @Override
        public int compare(Event e_1, Event e_2) {
            if (e_1.value > e_2.value) {
                return 1;
            }
            if (e_1.value < e_2.value) {
                return -1;
            }
            return 0;
        }
    }
}

File: PythonImplementation/LineSweep.py
from Event import Event
from heapq import *

from SortedListBasic import SortedListBasic

'''
class for linesweep algorithm
* Uses Sorted List
* Worst Case: O((n + k) * n)
* Best Case: O(n)
'''
class LineSweep:
    def __init__(self, edges):
        self.eventQueue = self.getEvents(edges)
        heapify(self.eventQueue)
        self.lineSweepStatus = SortedListBasic()
        self.intersections = set()

    def lineSweep(self):
        currEvent = self.eventQueue[0]
        currY = currEvent.point.y
        while self.eventQueue:
            currEvent = heappop(self.eventQueue)
            currY = currEvent.point.y
            if currEvent.category == 0:
                self.lineSweepStatus.add(currEvent.edge1, currY)
                predEdge = self.lineSweepStatus.predecessor(currEvent.edge1)
                sucEdge = self.lineSweepStatus.successor(currEvent.edge1)
                if predEdge:
                    intersection = predEdge.intersectionPoint(currEvent.edge1)
                    self.addIntersection(intersection, self.eventQueue,
                                         self.intersections, predEdge, currEvent.edge1)
                if sucEdge:
                    intersection = sucEdge.intersectionPoint(currEvent.edge1)
                    self.addIntersection(intersection, self.eventQueue,
                                         self.intersections, currEvent.edge1, sucEdge)

            elif currEvent.category == 1:
                predEdge = self.lineSweepStatus.predecessor(currEvent.edge1)
                sucEdge = self.lineSweepStatus.successor(currEvent.edge1)
                if predEdge and sucEdge:
                    intersection = predEdge.intersectionPoint(sucEdge)
                    self.addIntersection(intersection, self.eventQueue,
                                         self.intersections, predEdge, sucEdge)
                self.lineSweepStatus.remove(currEvent.edge1)

            else:
                self.lineSweepStatus.swapEdges(
                    currEvent.edge1, currEvent.edge2, currY)
                predEdge = self.lineSweepStatus.predecessor(currEvent.edge2)
                if predEdge:
                    intersection = predEdge.intersectionPoint(currEvent.edge2)
                    self.addIntersection(intersection, self.eventQueue,
                                         self.intersections, predEdge, currEvent.edge2)
                sucEdge = self.lineSweepStatus.successor(currEvent.edge1)
                if sucEdge:
                    intersection = currEvent.edge1.intersectionPoint(sucEdge)
                    self.addIntersection(intersection, self.eventQueue,
                                         self.intersections, currEvent.edge1, sucEdge)

        return self.intersections

    def addIntersection(self, intersection, eventQueue, intersections, edge1, edge2):
        if intersection:
            if intersection not in intersections:
                intersections.add(intersection)
                heappush(eventQueue, Event(2, intersection, edge1, edge2))

    def getXValue(self, segment, yVal):
        slope = (segment.p1.y - segment.p0.y) / (segment.p1.x - segment.p0.x)
        return ((yVal - segment.p1.y) / slope) + segment.p1.x

    def getEvents(self, edges):
        # Event category 0 is bottom, 1 is top, 2 is intersection
        events = []
        for edge in edges:
            if edge.p0.y > edge.p1.y:
                events.append(Event(1, edge.p0, edge))
                events.append(Event(0, edge.p1, edge))
            else:
                events.append(Event(0, edge.p0, edge))
                events.append(Event(1, edge.p1, edge))

        return events
